packages/ui-causal-graph-editor/src/shared/graph-layout/spring-layout.tsx

```diff
@@ -72,12 +79,121 @@ class SpringLayoutBuilder extends GraphLayoutBuilder<SpringLayout> {
+    tiers: GraphTiers,
+    graph: SimulationGraph,
+    orientation: DirectionType,
+    nodes: SimulationNodeWithGroup[]
+): void {
+    if (!Array.isArray(tiers)) {
+        const { order_nodes_by } = tiers;
+        if (order_nodes_by) {
+            const simNodes = graph.nodes();
+            const nodesOrder = getNodeOrder(simNodes, order_nodes_by, graph);
+            const sortedNodesOrderArray = Object.entries(nodesOrder)
+                .sort((a, b) => Number(a[1]) - Number(b[1]))
+                .map((entry) => entry[0]);
+
+            function forceOrder(): d3.Force<SimulationNodeWithGroup, undefined> {
+                function force(alpha: number): void {
+                    sortedNodesOrderArray.forEach((nodeName, index) => {
+                        const nodeSeparation = 200;
+                        const targetPosition = index * nodeSeparation;
+                        const targettedNode = nodes.find((obj) => obj.id === nodeName);
+
+                        if (targettedNode) {
+                            if (orientation === 'horizontal') {
+                                // Apply a nudge towards the target y position
+                                targettedNode.vy += (targetPosition - targettedNode.y) * alpha;
+                            } else {
+                                // Apply a nudge towards the target x position
+                                targettedNode.vx += (targetPosition - targettedNode.x) * alpha;
+                            }
+                        }
+                    });
+                }
+                return force;
+            }
+
+            // apply layer force
+            simulation.force('order', forceOrder());
+        }
+    }
+}
+
+/**
+ * Apply forces which are needed as part of a tiered layout.
+ * There is one force which snaps the nodes to the relevant layers and another force ordering the nodes within the layers.
+ *
+ * @param simulation the D3 simulation
+ * @param graph the simulation graph
+ * @param nodes array of the nodes
+ * @param tiers the tiers passed to the layout
+ * @param tiersSeparation the separation between the tiers
+ * @param orientation the orientation of the layout
+ */
+export function applyTierForces(
+    simulation: d3.Simulation<SimulationNode, D3SimulationEdge>,
+    graph: SimulationGraph,
+    nodes: SimulationNodeWithGroup[],
+    tiers: GraphTiers,
+    tiersSeparation: number,
+    orientation: DirectionType
+): void {
+    const tiersArray = getTiersArray(tiers, graph);
+
+    applyOrderNodesForce(simulation, tiers, graph, orientation, nodes);
+
+    function forceLayer(): d3.Force<SimulationNodeWithGroup, undefined> {
+        function force(alpha: number): void {
+            tiersArray.forEach((tier, index) => {
+                const targetPosition = index * tiersSeparation;
+                tier.forEach((nodeName) => {
+                    const targettedNode = nodes.find((obj) => obj.id === nodeName);
```

Comment:
This gets executed for every tick, right? For a small performance gain it might be useful to turn nodes into a mapping of {id -> node} outside of the force so this is just a lookup rather than a loop through the nodes list every time.

Same within the other force above

Reply:
Good point, have added this